Feature 3 : Gérer le profil

Neario opened the pull request:
close #3

Commit:
Merge pull request #29 from LangkaWS/develop

Develop

Commit:
liip imagine

ajout avec le composer du bundle liip/imagine

Commit:
Merge pull request #32 from LangkaWS/feature-2

Feature 2

Commit:
MyProfile

Création d'un formulaire , agissant sur le User entity
Creation d'une entité ManageProfile
gestion de la sécurité , être connecté avec le ["ROLE_USER"] pour accéder à la page profile

Commit:
gestion de la fonction profile

Supressin du controller manageprofile , pour fusionner la fonction profile dans le controller user

Commit:
Ajout d'une methode pseudo

Ajout de la methode pseudo

Commit:
Merge pull request #49 from LangkaWS/develop

Develop

Commit:
Ajout de la page d acuueil

Commit:
Html/Css

Mise en place de l html / css

Commit:
Update

Modidification du profil

Commit:
Addflash

Ajout du message de confirmation , manque le redirect .

LangkaWS commented on public/css/profile.css:
```diff
@@ -0,0 +1,25 @@
+.profile_wrap {
```
Tout ce qui est CSS on verra plus tard. Il y a Bootstrap qui gère pour le moment, le reste on centralisera dans un fichier main.css (pour tout ce qui est classes communes comme les titres des pages, etc.), et séparer après les trucs plus spécifiques.

LangkaWS commented on src/Controller/UserController.php:
```diff
@@ -3,6 +3,7 @@
 namespace App\Controller;
 
 use App\Entity\User;
+use App\Form\MyProfileType;
```
Plutôt utiliser un UserType pour garder la cohérence dans le naming.

LangkaWS commented on src/Controller/UserController.php:
```diff
@@ -57,4 +58,52 @@ public function register(
+        if($user->getIsActive()==0)
+            $user->setIsActive(false);
+        elseif($user->getIsActive()==1)
+            $user->setIsActive(true);
```
Tu peux enlever ce morceau de code :
Ton $user, tu fais un $this->getUser(). Donc les propriétés isAdmin et isActive proviennent de l'objet User, donc ont pour valeurs `true` ou `false`.

Neario replied:
fait

LangkaWS commented on src/Controller/UserController.php:
```diff
@@ -57,4 +58,52 @@ public function register(
+    public function myProfile(Request $request, EntityManagerInterface $em, UserPasswordEncoderInterface $encoder): Response
+    {
+        $user = $this->getUser();        
+        $profile = $this->createForm(MyProfileType::class, $user);
```
Plutôt préférer utiliser $form que $profile (uniformité du naming, par défaut Symfony l'appelle $form, comme ça dans ta méthode tu sais que tu parles du formulaire et pas du "profil")

Neario replied:
Ok

LangkaWS commented on src/Controller/UserController.php:
```diff
@@ -57,4 +58,52 @@ public function register(
+            $user->setFirstName($user->getFirstName());
+            $user->setPhone($user->getPhone());
+            $user->setEmail($user->getEmail());
+            $user->setPassword($hash);
```
Là je ne suis pas trop d'accord. Quand ton utilisateur va sur son profil, on lui demande son mdp (requis ; pour valider les changements je pense). Sauf que s'il se plante dans son mdp, ce nouveau mdp est enregistré en bdd et l'utilisateur a "changé" son mot de passe sans faire exprès.
Je préfèrerais laisser le mdp en non requis et si l'utilisateur ne le renseigne pas, on garde celui en bdd. Généralement sur les sites on ne demande pas le mdp pour éditer son profil.

LangkaWS commented on src/Controller/UserController.php:
```diff
@@ -57,4 +58,52 @@ public function register(
+            $user->setPassword($hash);
+            $user->setIsAdmin($user->getIsAdmin());
+            $user->setIsActive($user->getIsActive());
+            $user->setCampus($user->getCampus());
```
Sur toute cette partie, en fait ton code ne fait rien...
Pour résumer : 
- L'utilisateur arrive sur la page. Ton controller fait un getUser et montre les infos.
- L'utilisateur modifie ce qu'il a envie de modifier et soumet le formulaire. Par exemple il change son numéro de téléphone qui n'était pas renseigné (null) et met 0123456789.
- Le formulaire arrive dans ton controller avec les infos actualisées de $user.
- Il fait un `$user->getPhone()` : renvoie 0123456789
- Puis fait un `$user->setPhone($user->getPhone())` éq. `$user->setPhone(0123456789)` (or l'atrtibut phone est déjà égal à 0123456789
Tout ce que tu as besoin de faire c'est `$em->flush()`. D'ailleurs c'est comme ça quand tu génères automatiquement le CRUD avec Symfony ;)

Neario replied:
Compris

LangkaWS commented on src/Controller/UserController.php:
```diff
@@ -57,4 +58,52 @@ public function register(
+            $em->flush();
+
+            $this->addFlash('success', 'La modification du profil à bien été prise en compte.');
+        }
```
Là tu l'avais déjà dit, il manque un return

Neario replied:
return ajouter vers app_home

LangkaWS commented on src/Controller/UserController.php:
```diff
@@ -57,4 +58,52 @@ public function register(
+      * @return Response
+      * @Route("/profile", name="user_profile")
+      */
+    public function myProfile(Request $request, EntityManagerInterface $em, UserPasswordEncoderInterface $encoder): Response
```
Plutôt renommer la méthode en `edit`, pour correspondre au CRUD de Symfony (new, show, edit, delete)

Neario replied:
fait

LangkaWS commented on src/Form/MyProfileType.php:
```diff
@@ -0,0 +1,63 @@
+use Symfony\Component\Security\Core\Security;
+use Symfony\Component\OptionsResolver\OptionsResolver;
+
+class MyProfileType extends AbstractType
```
Renommer plutôt en UserType

Neario replied:
fait

LangkaWS commented on src/Form/MyProfileType.php:
```diff
@@ -0,0 +1,63 @@
+class MyProfileType extends AbstractType
+{
+    private $security;
+    public function __construct(Security $security){
```
Détail de mise en forme : mettre l'accolade à la ligne (uniformité du code)

Neario replied:
OK

LangkaWS commented on src/Form/MyProfileType.php:
```diff
@@ -0,0 +1,63 @@
+            ->add('password', RepeatedType::class, [
+                'type' => PasswordType::class,
+                'invalid_message' => 'The password fields must match.',
+                'options' => ['attr' => ['class' => 'password-field']],
```
Si la class n'est pas utilisée -> à enlever (n'est apparemment pas utilisé par Bootstrap ni dans ton CSS)

Neario replied:
Ok

LangkaWS commented on templates/base.html.twig:
```diff
@@ -6,6 +6,7 @@
         <title>{% block title %}Sortir.com{% endblock %}</title>
         <link href="https://cdn.jsdelivr.net/npm/bootstrap@5.0.0-beta1/dist/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-giJF6kkoqNQ00vy+HMDP7azOuL0xtbfIcaT9wjKHr8RbDVddVHyTfAAsrekwKmP1" crossorigin="anonymous">
         <link href="{{ asset('css/main.css')}}" rel="stylesheet">
+        <link href="{{ asset('css/profile.css')}}" rel="stylesheet">
```
cf commentaire css

Neario replied:
Css viré pour le moment

LangkaWS commented on templates/user/myProfile.html.twig:
```diff
@@ -0,0 +1,43 @@
+{% extends 'base.html.twig' %} 
```
Renommer en show.html.twig (uniformité du nommage, nom automatique donné par Symfony quand tu fais un CRUD).
De plus, ce fichier est utilisé pour afficher un profil utilisateur, que ce soit le sien ou celui d'un autre ; ça évitera de devoir faire 2 fichiers html qui seront quasi identiques.
Il faut juste rajouter des blocs if selon si l'utilisateur est sur sa page ou non.

Neario replied:
CRUD activé , nommage modifié

LangkaWS commented on templates/user/myProfile.html.twig:
```diff
@@ -0,0 +1,43 @@
+<div class="profile_wrap">
+    <h3>Mon profil</h3>
+    {{ form_start(profile) }}
+    <div class="rows">
+        {{ form_row(profile.username) }}
+    </div>
+
+    <div class="rows">
+        {{ form_row(profile.lastName) }}
+    </div>
+
+    <div class="rows">
+        {{ form_row(profile.firstName) }}
+    </div>
+
+    <div class="rows">
+        {{ form_row(profile.phone) }}
+    </div>
+
+    <div class="rows">
+        {{ form_row(profile.email) }}
+    </div>
+
+    <div class="rows">
+        {{ form_row(profile.password) }}
+    </div>
+     <div class="rows">
+        {{ form_row(profile.campus) }}
+    </div>
```
En utilisant `{{ form_widget(profile) }}` Symfony te génère tout seul les div ;) ça fait moins de lignes de codes à faire :)
La mise en page tu pourras la faire avec le CSS.

Neario replied:
suppression des divs inutile

LangkaWS commented on templates/user/myProfile.html.twig:
```diff
@@ -0,0 +1,43 @@
+{% block title %}My profile ! {% endblock %} 
+
+{% block main %}
+<div class="profile_wrap">
```
Ta div ici ne sert à rien car tu te retrouves avec le code :
```
<main>
    <div class="profile_wrap">
        <h3>Mon profil</h3>
        <form>...</form>
    </div>
</main>
```
Pour la mise en forme, tu peux regarder ce que Symfony crée avec le CRUD, il fait une table, ça t'évite les display: flex, dans tous les sens dans le CSS qui ne flexent même pas :/

Neario replied:
sisi , ça flex en column , afin d'aligner avec align-items au centre de ma page

LangkaWS commented on templates/user/myProfile.html.twig:
```diff
@@ -0,0 +1,43 @@
+    <div class="button">
+        <button type="submit" class="btn btn-primary">Enregistrer</button>
+        <button type="reset" class="btn btn-secondary">Effacer</button>
+        <p>{{app.user.campus}}</p>
```
Un petit campus qui s'est perdu par là ^^

Neario replied:
corrigé c'était un test

LangkaWS reviewed (changes requested):
Pas mal de choses à revoir, ce sont des suggestions, à toi de voir :)

Commit:
Ajout du get

Ajout de la method get et post avec {username]

Commit:
Redirect

ajout du redirect

Commit:
Mise a jour du code

Commit:
nommage function/twig

Nommage plus clair pour la function edit et de son fichier twig

LangkaWS commented on src/Controller/UserController.php:
```diff
@@ -57,4 +58,32 @@ public function register(
+            return $this->redirectToRoute("app_home");
+        }
+        return $this->render('user/edit.html.twig', [
+            'controller_name' => 'ManageProfileController',
```
le controller_name si tu n'en as pas besoin, tu peux l'enlever :)

Neario replied:
fait

LangkaWS commented on templates/user/edit.html.twig:
```diff
@@ -0,0 +1,34 @@
+    <h3>Mon profil</h3>
+    {{ form_start(profile) }}
+    
+        {{ form_row(profile.username) }}  
+    
+        {{ form_row(profile.lastName) }}
+    
+        {{ form_row(profile.firstName) }}
+    
+        {{ form_row(profile.phone) }}
+    
+        {{ form_row(profile.email) }}
+    
+        {{ form_row(profile.password) }}
+
+        {{ form_row(profile.campus) }}
```
Tout ça équivaut à {{ form_widget(profile) }} en fait ^^ ça économise 10 lignes :D

Neario replied:
OK

Neario replied:
Oui exact

LangkaWS reviewed (changes requested):
Attention, quand j'édite et j'enregistre sans mettre de mot de passe, j'ai l'erreur suivante : 
"Expected argument of type "string", "null" given at property path "password"."

Commit:
Mode Crud Activé

J'ai refait avec le Crud activé , drop le css , activé le from_thème

Commit:
Update security.yaml

Commit:
Update edit.html.twig

J'ai regroupé les rows en un widget

Commit:
Delete profile.css

Heroponne commented on src/Controller/UserController.php:
```diff
@@ -57,4 +52,84 @@ public function register(
             'registerForm' => $registerForm->createView()
         ]);
     }
+
+    /**
+     * @Route("/user", name="user_index", methods={"GET"})
+     */
+    public function index(UserRepository $userRepository): Response
```
Peut-être juste mettre / en route, et pas /user sinon l'url sera /user/user

Neario replied:
fait

Heroponne commented on templates/user/index.html.twig:
```diff
@@ -0,0 +1,53 @@
+{% extends 'base.html.twig' %}
+
+{% block title %}User index{% endblock %}
+
+{% block body %}
+    <h1>User index</h1>
+
```
on n'en a pas encore besoin pour le moment (peut-être la partie admin), ais bon, juste chenger le block body par main et le title

Commit:
Petit ajustement formulaire

emplacement des champs réadapter selon la maquette

Commit:
Update UserController.php

Petit ajout de code pour que seulement la bonne personne puisse modifier son profil

Commit:
Modif password

Modif avec et sans password

LangkaWS commented on src/Controller/UserController.php:
```diff
@@ -57,4 +52,91 @@ public function register(
+        ]);
+    }
+
+    /**
+     * @Route("/new", name="user_new", methods={"GET","POST"})
+     */
+    public function new(Request $request): Response
+    {
+        $user = new User();
+        $form = $this->createForm(UserType::class, $user);
+        $form->handleRequest($request);
+
+        if ($form->isSubmitted() && $form->isValid()) {
+            $entityManager = $this->getDoctrine()->getManager();
+            $entityManager->persist($user);
+            $entityManager->flush();
+
+            return $this->redirectToRoute('user_index');
+        }
+
+        return $this->render('user/new.html.twig', [
+            'user' => $user,
+            'form' => $form->createView(),
+        ]);
+    }
```
A retirer car les nouveaux utilisateurs sont créés via la méthode register

Neario replied:
Viré

Heroponne commented on src/Entity/User.php:
```diff
@@ -117,9 +119,13 @@ public function setEmail(string $email): self
      */
     public function getUsername(): string
     {
-        return (string) $this->username;
+        return (string) $this->email;
```
Le souci n'avait pas été réglé ?  c'est normal que le getusername renvoie encore l'email ?

LangkaWS replied:
C'est parce qu'il lui manque le bugfix je pense.

Neario replied:
Fait

LangkaWS commented on src/Controller/UserController.php:
```diff
@@ -57,4 +52,91 @@ public function register(
+    }
+
+    /**
+     * @Route("/{id}", name="user_show", methods={"GET"})
```
Tu pourrais faire comme pour edit, utiliser le {username} plutôt que l'id :)

Neario replied:
fait

Heroponne commented on src/Form/UserType.php:
```diff
@@ -0,0 +1,66 @@
+                'data' => $data->getLastName(),
+                ])
+            ->add('phone', TextType::class,[
+                'label' => 'Téléphone : ',
+                'data' => $data->getPhone(),
+                ])
```
mettre un 'required' => false

Neario replied:
fait

LangkaWS commented on src/Controller/UserController.php:
```diff
@@ -57,4 +52,91 @@ public function register(
+     */
+    public function edit(Request $request, User $user, UserPasswordEncoderInterface $encoder): Response
+    {
+        if($user->getEmail() != $this->getUser()->getEmail()){
+            return $this->redirectToRoute('app_home');
+        }
```
La condition sert à vérifier si l'utilisateur essaie de modifier le profil de quelqu'un d'autre ?
Dans ce cas, c'est peut-être plus indiqué d'enlever {username} du path, et à chaque fois que tu dois référer au user, tu fais un $this->getUser(), à voir.

LangkaWS commented on src/Entity/User.php:
```diff
@@ -117,9 +119,13 @@ public function setEmail(string $email): self
      */
     public function getUsername(): string
     {
-        return (string) $this->username;
+        return (string) $this->email;
```
Non, là il faut que tu pull develop pour avoir le fix.

Neario replied:
Fix pull

LangkaWS commented on src/Entity/User.php:
```diff
@@ -117,9 +119,13 @@ public function setEmail(string $email): self
+        return (string) $this->email;
     }
 
+    public function getPseudo(): string
+    {
+        return (String) $this->username;
+    }
```
Et tu peux retirer celle-là une fois que tu as re pull et merge develop dans ta feature.

Neario replied:
fix fait

LangkaWS commented on src/Form/UserType.php:
```diff
@@ -0,0 +1,66 @@
+                'label' => 'Nom : ',
+                'data' => $data->getLastName(),
+                ])
+            ->add('phone', TextType::class,[
+                'label' => 'Téléphone : ',
+                'data' => $data->getPhone(),
+                ])
```
Sur le phone, j'ai un message d'erreur comme quoi il est requis.
Le phone n'est pas requis en bdd, donc pas besoin d'avoir cette restriction :)

Neario replied:
Corrigé

Commit:
phone fix

phone peut être null

LangkaWS commented on templates/user/edit.html.twig:
```diff
@@ -0,0 +1,26 @@
+{% form_theme form 'bootstrap_4_layout.html.twig' %}
+    <h3>Edit User</h3>
+
+    {{ form_start(form) }}
+    
+    {{ form_widget(form) }}
+
+    <div class="button mt-4">
+        <button type="submit" class="btn btn-primary">Enregistrer</button>
+        <button type="reset" class="btn btn-secondary">Effacer</button>
+    </div>
+
+    {{ form_end(form) }}
+</div>
+
+    {#{{ include('user/_form.html.twig', {'button_label': 'Update'}) }}#}
```
Ici tu peux utiliser le include _form qui est donné par défaut, ça ne marche pas ?
Au passage tu as une balise </div> mais pas de balise ouvrante ^^

Neario replied:
le bonton update inclus ne fonctionne pas

LangkaWS replied:
Ah bizarre j'avais testé ça avait bien fonctionné

LangkaWS commented on templates/user/new.html.twig:
```diff
@@ -0,0 +1,11 @@
+{% extends 'base.html.twig' %}
+
+{% block title %}New User{% endblock %}
+
+{% block body %}
+    <h1>Create new User</h1>
+
+    {{ include('user/_form.html.twig') }}
+
+    <a href="{{ path('user_index') }}">back to list</a>
+{% endblock %}
```
Comme l'inscription se passe sur register.html.twig, tu peux supprimer ce fichier

Neario replied:
viré

LangkaWS commented on templates/user/index.html.twig:
```diff
@@ -0,0 +1,53 @@
+                <th>Password</th>
+                <th>Username</th>
+                <th>LastName</th>
+                <th>FirstName</th>
+                <th>Phone</th>
+                <th>IsAdmin</th>
+                <th>IsActive</th>
+                <th>ProfilePicture</th>
+                <th>actions</th>
+            </tr>
+        </thead>
+        <tbody>
+        {% for user in users %}
+            <tr>
+                <td>{{ user.id }}</td>
+                <td>{{ user.email }}</td>
+                <td>{{ user.roles ? user.roles|json_encode : '' }}</td>
+                <td>{{ user.password }}</td>
+                <td>{{ user.username }}</td>
+                <td>{{ user.lastName }}</td>
+                <td>{{ user.firstName }}</td>
+                <td>{{ user.phone }}</td>
+                <td>{{ user.isAdmin ? 'Yes' : 'No' }}</td>
+                <td>{{ user.isActive ? 'Yes' : 'No' }}</td>
+                <td>{{ user.profilePicture }}</td>
+                <td>
+                    <a href="{{ path('user_show', {'id': user.id}) }}">show</a>
+                    <a href="{{ path('user_edit', {'id': user.id}) }}">edit</a>
+                </td>
+            </tr>
+        {% else %}
+            <tr>
+                <td colspan="12">no records found</td>
+            </tr>
+        {% endfor %}
+        </tbody>
+    </table>
+
+    <a href="{{ path('user_new') }}">Create new</a>
+{% endblock %}
```
Comme on n'affichera pas la liste des utilisateurs, tu peux supprimer ce fichier aussi

Neario replied:
supprimé

Commit:
Merge branch 'develop' into feature-3

LangkaWS reviewed (changes requested):
A rajouter peut-être après avoir merge une première fois : 
Quand l'utilisateur clique sur "Mon profil" il voit son profil et il doit appuyer sur "modifier" pour pouvoir modifier les champs ; ça éviterait les modifications accidentelles. - A noter que c'est pas demandé dans le PDF des cas d'utilisation ni dans le description produit

Commit:
Nouvelle version

Avec le profil utilisateur , un bouton edit, le bug fix .

Commit:
Update show.html.twig

Commit:
Update show.html.twig

update avec la condition user == app.user

Commit:
Update edit.html.twig

inclusion du forlulaire fournis par le crud

LangkaWS reviewed (approved):
C'est bon j'ai plus de commentaires :D

Heroponne reviewed (approved):
ok pour moi :)
